src/main.c

```diff
@@ -351,13 +356,75 @@ key_binding_func_key (MBWindowManager   *wm,
   hd_dbus_send_event (s);
 }
 
+/* Fill in the array with input devices type is touchscreen */
+static void
+enumerate_input_devices (Display *dpy)
+{
+  static Atom atom_touchscreen = None;
+
+  if (atom_touchscreen == None)
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen != None)
+    {
+      XDeviceInfo *devinfo;
+      int i, ndev, nclass = 0;
+      XEventClass eclass[64];
+
+      /* get XInput DeviceMotion events */
+      devinfo = XListInputDevices (dpy, &ndev);
+
+      if (ndev > G_N_ELEMENTS(eclass))
+        ndev = G_N_ELEMENTS(eclass);
+
+      if (ndev > G_N_ELEMENTS(xdevice_is_ts))
+        ndev = G_N_ELEMENTS(xdevice_is_ts);
+
+      for (i = 0; i < ndev; i++)
+        {
+          XDeviceInfo info = devinfo[i];
+
+          if (info.use == IsXExtensionPointer)
+            {
+              XDevice *dev = XOpenDevice (dpy, info.id);
```

Comment:
Where is the call to XCloseDevice()?

Reply:
'The XCloseDevice request makes an input device inaccessible to a client through input extension protocol requests'. So, IIUC we should not XCloseDevice() if we want events to come. However, you have a point here - I think before re-enumerating, currently opened devices shall be closed.

Reply:
According to manual it should also be done before program terminating:
```
Before terminating, and client that has opened input devices through
the input extension should close them via CloseDevice.
```

Reply:
Right, but I am not sure this is achievable for h-d. Will see what can be done though.